### PR TITLE
feat: Added tax to order checkout page

### LINKED ISSUE #2799: Tax Info: Show tax in event ticket section and checkout information
Please show the tax percentage and the tax sum in the ticket section of the event page and the checkout information (if tax info has been added) as well as in the invoice PDF.

* Show 
a) a subtotal
b) tax and 
c) Grand Total

Screenshot from 2019-04-25 00-24-04

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Screenshot 2019-05-18 at 10 24 13 PM

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [X] My branch is up-to-date with the Upstream `development` branch.
- [X] The acceptance, integration, unit tests and linter pass locally with my changes 
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

#### Short description of what this resolves:
- Shows tax info at the checkout, order-confirmation and order-placed page
- Notifies user of the tax status related to his order

#### Changes proposed in this pull request:

- Calculate Tax and show it as the Grand Total
- Communicate the same data back to the server

Fixes #2799

### CODE REVIEW

**Comment by mrsaicharan1** [comment]
Tax should be added to the discounted price right? @shreyanshdwivedi @uds5501

**Comment by mrsaicharan1** [comment]
@uds5501 @shreyanshdwivedi I've added the tax to the order-summary. To calculate order total, I've used something like this in the `component` file  `order-summary.js`
`  taxed_price: computed('data.amount', 'event.tax.rate', function(){
    return this.get('data.amount') * this.get('event.tax.rate') / 100 + this.get('data.amount');
  })`
.  But when I access this `taxed_price`, it shows nothing. What's happening here?

**Comment by uds5501** [comment]
@mrsaicharan1 I will pull your PR and see what's going wrong

**Comment by mrsaicharan1** [comment]
@uds5501 Thanks for that. Just some info, the tax is properly shown under `Tax`, 
```
   <td>{{event.tax.rate}}%</td>

```when you comment out the code for 
```
              {{#if not event.tax.isTaxIncludedInPrice}}
                {{currency-symbol eventCurrency}} {{number-format data.amount}}
              {{else}}
                {{currency-symbol eventCurrency}} {{number-format taxed_price}}
              {{/if}}
```

**Comment by uds5501** [comment]
@mrsaicharan1 Even if I don't comment the code, I am getting this :
image
So, I don't think that implementation's wrong. Just see that if there is no tax info, display a 0%

Also, you may want to fix what data is being sent back to server, if the ticket amount was 1500, it's still being sent 1500 instead of 1650 :moneybag:

File: app/routes/orders/view.js [hunk 1]
```diff
@@ -11,10 +11,13 @@ export default Route.extend({
...
-    const eventDetails = await order.query('event', {});
+    const eventDetails = await order.query('event', {
+      include: 'tax'
+    });
     return {
       order,
-      form: await eventDetails.query('customForms', {
+      event : eventDetails,
+      form  : await eventDetails.query('customForms', {
         'page[size]' : 50,
```
Comment (uds5501) [inline app/routes/orders/view.js]: @mrsaicharan1 I am using : 
```js

const customEventDetails = await order.query('event', {include : 'tax'});
    return {
      order,
      event : customEventDetails,
      form: await eventDetails.query('customForms', {
        'page[size]' : 50,
        sort         : 'id'
      })
    };
```
see if it works for you if it hasn't been for now

**Comment by mrsaicharan1** [comment]
@uds5501 Yeah, might be some error on my side. I was wondering what was going wrong. Thanks for this. 
I'll add a check for tax absence + fix the data being sent back to the server

File: app/routes/orders/view.js (hunk 1 above)
Comment (mrsaicharan1) [reply app/routes/orders/view.js]: Yes, believe that creating a separate `customEventDetails` is more appropriate.

**Comment by mariobehling** [comment]
In the tax form in wizard step one there are two options:
1. Add tax on top (US Way)
2. Include tax (EU/Asia Way)

Could you please verify that you adhere to the chosen option?

**Comment by mrsaicharan1** [comment]
> In the tax form in wizard step one there are two options:
> 
> 1. Add tax on top (US Way)
> 2. Include tax (EU/Asia Way)
> 
> Could you please verify that you adhere to the chosen option?

Yes, I've added checks for this. If the tax is included, will show the total amount directly. If the tax is not included, will calculate price+tax and show `Grand Total`

**Comment by mrsaicharan1** [comment]
I'm doing something like this:

`app/components/orders/order-summary.js`

```
async didInsertElement() {
    let discountCode = await this.get('data.discountCode');
    let tickets = await this.get('data.tickets');
    tickets.forEach(ticket => {
      ticket.set('discount', 0);
    });
    if (!event.tax.isTaxIncludedInPrice) {
      let order = await this.get('data');
      order.set('amount', this.get('taxed_price'));
    }
```
I've used this but the order amount being sent to the server still remains the same. @shreyanshdwivedi @uds5501

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 @shreyanshdwivedi @uds5501  I've tried but I couldn't figure out how to send the `taxed_price` to the server when the `tax` is not included in the price. How should this be handled?

**Comment by uds5501** [comment]
@mrsaicharan1 Try modifying the amount data in `components/forms/orders/order-form.js`

I think you can modify this function to make the data save work
```js
    submit(data) {
      this.onValid(() => {
        this.sendAction('save', data);
      });
    },

```

**Comment by mrsaicharan1** [comment]
> @mrsaicharan1 Try modifying the amount data in `components/forms/orders/order-form.js`
> 
> I think you can modify this function to make the data save work
> 
> ```js
>     submit(data) {
>       this.onValid(() => {
>         this.sendAction('save', data);
>       });
>     },
> ```

How is `data` being accessed here?

**Comment by uds5501** [comment]
@mrsaicharan1 you have to add the `taxed_price` element over here too. You have the amount already, see if you can include event data along with it and implement the taxed_price in this route.

**Comment by pradeepgangwar** [comment]
@mrsaicharan1 I pulled your PR locally and checked it out. The FE section looks fine. Were you able to figure out "how to send data back to server"? Let me know any other issue you are facing, I will chec k the PR once more.

**Comment by mrsaicharan1** [comment]
@pradeepgangwar I was just fixing up some priority issues over on the server. Will update this PR in a few hours, as soon as I'm done with that!

**Comment by pradeepgangwar** [comment]
I see the possibility of modifying `data.amount` directly. Including a new `taxedPrice` variable seems redundant if we are able to make changes directly to data.amount, that will automatically send new amount to server.

> I was just fixing up some priority issues over on the server. Will update this PR in a few hours, as soon as I'm done with that!

That's fine :).

File: app/templates/components/orders/order-summary.hbs [hunk 2]
```diff
@@ -55,6 +56,13 @@
             <td>
               {{currency-symbol eventCurrency}} {{number-format (mult (sub ticket.price ticket.discount) (ticket-attendees data.attendees ticket.attendees))}}
             </td>
+            <td>
+              {{#if event.tax.isTaxIncludedInPrice}}
+                {{t 'Included'}}
```
Comment (pradeepgangwar) [inline app/templates/components/orders/order-summary.hbs]: Change this to : **The price is inclusive of all taxes**.

**Comment by mrsaicharan1** [comment]
> I see the possibility of modifying `data.amount` directly. Including a new `taxedPrice` variable seems redundant if we are able to make changes directly to data.amount, that will automatically send new amount to server.
> 
> > I was just fixing up some priority issues over on the server. Will update this PR in a few hours, as soon as I'm done with that!
> 
> That's fine :).

I've set the order amount just after the `price+tax` had been calculated. But the server data still shows up to be the old data.

File: app/components/orders/order-summary.js [hunk 3]
```diff
@@ -14,6 +14,13 @@ export default Component.extend(FormMixin, {
     );
   }),
 
+  taxed_price: computed('data.amount', 'event.tax.rate', function() {
```
Comment (mrsaicharan1) [inline app/components/orders/order-summary.js]: @pradeepgangwar I've set the order amount just after the price+tax had been calculated. But the server data still shows up to be the old data.

File: app/components/orders/order-summary.js [hunk 4]
```diff
@@ -14,6 +14,13 @@ export default Component.extend(FormMixin, {
...
+    let taxedPrice = this.get('data.amount') * this.get('event.tax.rate') / 100 + this.get('data.amount');
+    let order = this.get('data');
+    order.set('amount', taxedPrice);
+    return taxedPrice;
```
Comment (mrsaicharan1) [inline app/components/orders/order-summary.js]: @pradeepgangwar I've set the order amount just after the price+tax had been calculated. But the server data still shows up to be the old data.

Comment (pradeepgangwar) [reply app/components/orders/order-summary.js]: Checking it.

File: app/components/orders/order-summary.js [hunk 5]
```diff
@@ -14,6 +14,13 @@ export default Component.extend(FormMixin, {
...
+  }),
+
```
Comment (pradeepgangwar) [inline app/components/orders/order-summary.js]: Remove any changes in this file.

File: app/routes/orders/new.js [hunk 6]
```diff
@@ -27,11 +27,14 @@ export default Route.extend({
       });
     });
 
-    const eventDetails = await order.query('event', {});
+    const eventDetails = await order.query('event', {
+      include: 'tax'
+    });
```
Comment (pradeepgangwar) [inline app/routes/orders/new.js]: No changes needed here.

File: app/routes/orders/view.js [hunk 7]
```diff
@@ -11,10 +11,13 @@ export default Route.extend({
       include : 'attendees,tickets,event',
       reload  : true
     });
-    const eventDetails = await order.query('event', {});
+    const eventDetails = await order.query('event', {
+      include: 'tax'
+    });
```
Comment (pradeepgangwar) [inline app/routes/orders/view.js]: No changes needed here.

File: app/templates/components/orders/order-summary.hbs [hunk 8]
```diff
@@ -65,12 +73,16 @@
...
             </div>
           </th>
           <th colspan="4">
             <div class="ui aligned small primary icon">
-              {{currency-symbol eventCurrency}} {{number-format data.amount}}
+              {{#if event.tax.isTaxIncludedInPrice}}
+                {{currency-symbol eventCurrency}} {{number-format data.amount}}
+              {{else}}
+                {{currency-symbol eventCurrency}} {{number-format taxed_price}}
+              {{/if}}
```
Comment (pradeepgangwar) [inline app/templates/components/orders/order-summary.hbs]: No change here. `{{currency-symbol eventCurrency}} {{number-format data.amount}}` Is sufficient now. :)

**Comment by mrsaicharan1** [comment]
> @mrsaicharan1 I tried it and found a better approach to tackle this. You may be getting error probably because you were trying to modify the object passed to component not the model itself. We need modifications in model to propagate them to server.
> 
> 1. Since we know the taxes beforehand what we can do is to add the taxes to order amount beforehand instead of doing it on checkout stage (order-summary component).
> 
> ```diff
> diff --git a/app/controllers/public/index.js b/app/controllers/public/index.js
> index c1b1d271..bda7a22e 100644
> --- a/app/controllers/public/index.js
> +++ b/app/controllers/public/index.js
> @@ -112,8 +112,13 @@ export default Controller.extend({
>          this.set('isLoginModalOpen', true);
>          return;
>        }
> -      let order = this.get('model.order');
> -      let event = order.get('event');
> +      let order = await this.get('model.order');
> +      let event = await this.get('model.event');
> +      let tax = await this.get('model.event.tax');
> +      if (!tax.taxIncludedInPrice) {
> +        let taxedPrice = (order.amount) * (tax.rate) / 100 + (order.amount);
> +        order.set('amount', taxedPrice);
> +      }
>        order.tickets.forEach(ticket => {
>          let numberOfAttendees = ticket.orderQuantity;
>          while (numberOfAttendees--) {
> ```
> 
> 1. While trying your approach when I tried to alter the model.order.amount in order-summary and proceeded to save the changes I got 403 error stating that I cannot change the order amount. This is actually true because we do not want anyone with user privileges to alter the order details at the checkout step.
> 
> Also please make changes to `routes/orders/placed.js` similar to `routes/orders/view.js` and `routes/orders/new.js`
> 
> Let me know if you have any doubt.

Acknowledged. Please check the changes now.

**Comment by mrsaicharan1** [comment]
@pradeepgangwar Thanks for the clarification. Yes, updating the model at the checkout page is forbidden because of the obvious privileges constraint. I think this is good to go now.

**Comment by mrsaicharan1** [comment]
@pradeepgangwar Thanks for the clarification. Yes, updating the model at the checkout page is forbidden because of the obvious privileges constraint. I think this is good to go now.

**Comment by mrsaicharan1** [comment]
> This looks good @mrsaicharan1 :D.
> 
> Can you recheck if this is working fine. I think we need to add this too.
> 
> ```diff
> diff --git a/app/routes/public.js b/app/routes/public.js
> index 1739bb2c..291e2f83 100644
> --- a/app/routes/public.js
> +++ b/app/routes/public.js
> @@ -8,7 +8,7 @@ export default Route.extend({
> 
>    model(params) {
>      return this.store.findRecord('event', params.event_id, {
> -      include: 'social-links,event-copyright,speakers-call'
> +      include: 'social-links,event-copyright,speakers-call,tax'
>      });
>    }
>  });
> ```

I had the same doubt. But it seems to be working fine.

File: app/controllers/public/index.js [hunk 9]
```diff
@@ -112,8 +112,13 @@ export default Controller.extend({
         this.set('isLoginModalOpen', true);
         return;
       }
-      let order = this.get('model.order');
-      let event = order.get('event');
+      let order = await this.get('model.order');
+      let event = await order.get('event');
+      let tax = this.get('model.event.tax');
```
Comment (mrsaicharan1) [inline app/controllers/public/index.js]: @pradeepgangwar I removed `await` as we are including `tax` while fetching the event model.

Comment (mrsaicharan1) [reply app/controllers/public/index.js]: Also, it's working fine with this.

File: app/controllers/public/index.js [hunk 10]
```diff
@@ -112,8 +112,13 @@ export default Controller.extend({
         this.set('isLoginModalOpen', true);
         return;
       }
-      let order = this.get('model.order');
-      let event = order.get('event');
+      let order = await this.get('model.order');
```
Comment (CosmicCoder96) [inline app/controllers/public/index.js]: use ES6 getters everywhere.

Comment (mrsaicharan1) [reply app/controllers/public/index.js]: @CosmicCoder96 Done.

File: app/controllers/public/index.js [hunk 11]
```diff
@@ -112,8 +112,13 @@ export default Controller.extend({
         this.set('isLoginModalOpen', true);
         return;
       }
-      let order = this.get('model.order');
-      let event = order.get('event');
+      let order = await this.model.order;
+      let event = order.event;
```
[the remark this answers is not in the document]
Comment (pradeepgangwar) [reply app/controllers/public/index.js]: @mrsaicharan1 `use let { event } = order`

File: app/controllers/public/index.js [hunk 12]
```diff
@@ -112,8 +112,13 @@ export default Controller.extend({
...
+      let tax = event.tax;
```
[the remark this answers is not in the document]
Comment (pradeepgangwar) [reply app/controllers/public/index.js]: let { tax } = event

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 Just check this out whenever you get free.

**Comment by kushthedude** [comment]
@pradeepgangwar @mrsaicharan1 @uds5501 
image
Alignment Issue , Tax is not getting add in final amount

**Pull request merged**